### DISCUSSION TITLE
OData response format

### DISCUSSION THREAD

**Discussion by rbalagangadharan** [OData/AspNetCoreOData#1217] [Q&A]
Hi All,

I am just looking into this [ODataRoutingSample](https://github.com/OData/AspNetCoreOData/blob/main/sample/ODataRoutingSample/Controllers/AccountsController.cs),

When I browse this, I am getting below response.
```
[
    {
        "Name": "Freezing"
    },
    {
        "Name": "Bracing"
    },
    {
        "Name": "Chilly"
    },
    {
        "Name": "Cool"
    },
    {
        "Name": "Mild"
    },
    {
        "Name": "Warm"
    },
    {
        "Name": "Balmy"
    },
    {
        "Name": "Hot"
    },
    {
        "Name": "Sweltering"
    },
    {
        "Name": "Scorching"
    }
]
```

I knew these end points are Non-OData Endpoint Mapping. Is there a way to create a response same like OData Endpoint mapping without registering in startup, like below

```
{
	"@odata.context": "http://localhost/odata/account/$metadata",
	"@odata.count": 10,
	"value": [
		{
			"Name": "Freezing"
		},
		{
			"Name": "Bracing"
		},
		{
			"Name": "Chilly"
		},
		{
			"Name": "Cool"
		},
		{
			"Name": "Mild"
		},
		{
			"Name": "Warm"
		},
		{
			"Name": "Balmy"
		},
		{
			"Name": "Hot"
		},
		{
			"Name": "Sweltering"
		},
		{
			"Name": "Scorching"
		}
	],
	"@odata.nextLink": "http://localhost/odata/account?$skip=10"
}
```

**Comment by julealgon** [comment]
I'm pretty sure you _need_ a backing EDM model that matches your route to get the OData format on the responses, so the answer would be no: you need to define the EDM in the startup.

Someone on the team can confirm if there is no other way as I'm not 100% sure about that.

Is there a reason you don't want to define the EDM?

**Reply by rbalagangadharan** [reply]
Yeah, in my case the EDM is dynamic and will be generated at run time.

**Reply by julealgon** [reply]
So wait.... you _have_ an EDM? If you have one, even if it is generated dynamically, you can still modify it on a per-request basis even when using `[EnableQuery]`. There was an issue about that not too long ago I'm pretty sure.

Perhaps you can share a bit more detail about your scenario?

**Reply by rbalagangadharan** [reply]
@julealgon We don't have pre-defined entities, so we can't register the EDM with route like below
`builder.Services.AddControllers().AddOData(
    options => options.Select().Filter().OrderBy().Expand().Count().SetMaxTop(null).AddRouteComponents(
        "odata",
        modelBuilder.GetEdmModel()));`
In our case the Entity will be defined at runtime. `[EnableQuery]` works well, but the response we are getting is not OData compatible (as mentioned above)

**Reply by julealgon** [reply]
> In our case the Entity will be defined at runtime.

When exactly?

**Reply by rbalagangadharan** [reply]
on user request. 
```
public class DynamicController : ODataController
{
    private readonly MyDynamicDbContext _dbContext;
    private readonly DynamicStore _dynamicStore;

    public DynamicController(MyDynamicDbContext dbContext, DynamicStore dynamicStore)
    {
        _dbContext = dbContext;
        _dynamicStore = dynamicStore;
    }
	
    [HttpGet("odata/{dynamicEntity}")]
    [EnableQuery(PageSize = 100)]
    public IActionResult Get(string dynamicEntity)
    {
        var type = _dynamicStore.GenerateDynamic(dynamicEntity); // here we create class using Reflection.
        var results = _dbContext.EntitySet(type); // Here we invoke the Set method
        return Ok(results);
    }
}

```

When user request https://localhost:8282/odata/Test, the response is coming as json format, not as OData compatible.

**Comment by julealgon** [comment]
@rbalagangadharan this might help.
- https://github.com/OData/AspNetCoreOData/issues/381

If you are able to generate an EDM model on the spot based on that request, you should be able to dynamically set the EDM using that `ODataFeatures().Model` code there.

If that doesn't work, then I'm unfortunately out of ideas and you'll have to wait for someone on the team to help you out.
